src/orion/service/broker/local.py

```diff
@@ -0,0 +1,198 @@
+                space=client._experiment.space.configuration,
+                version=client._experiment.version,
+                mode="x",
+                max_trials=client._experiment.max_trials,
+                working_dir=client._experiment.working_dir,
+                metadata=client._experiment.metadata,
+            )
+        )
+
+    def suggest(self, request: RequestContext) -> Dict:
+
+        # NOTE: need to fix lost trials here
+        # so reschedule them
+        experiment_name = request.data.pop("experiment_name")
+        client = retrieve_experiment_client(request, experiment_name)
+
+        if client.is_broken:
+            raise BrokenExperiment()
+
+        trial = client.suggest(**request.data)
+
+        small_trial = self._make_trial(trial)
+        return success(dict(trials=[small_trial]))
+
+    def _make_trial(self, trial):
+        return RemoteTrial.make_json_from_trial(trial)
+
+    def observe(self, request: RequestContext) -> Dict:
+        experiment_name = request.data.pop("experiment_name")
+        client = retrieve_experiment_client(request, experiment_name)
+        storage = client.storage
+
+        trial_hash = request.data.pop("trial_hash")
+        results = request.data.get("results")
+
+        # pylint: disable=protected-access
+        trial = storage.get_trial(uid=trial_hash, experiment_uid=client._experiment.id)
+
+        assert trial is not None, "Trial not found"
+        client.observe(trial, results)
+
+        return success({})
+
+    def is_done(self, request: RequestContext) -> Dict:
+        experiment_name = request.data.pop("experiment_name")
+        client = retrieve_experiment_client(request, experiment_name)
+
+        if client.is_broken:
+            raise BrokenExperiment()
+
+        return success(dict(is_done=client.is_done))
+
+    def heartbeat(self, request: RequestContext) -> Dict:
+        storage = get_storage_for_user(request)
+        trial_id = request.data.get("trial_id")
+
+        # pylint: disable=protected-access
+        results = storage._db._db["trials"].update_one(
+            {
+                "_id": ObjectId(trial_id),
+                "status": "reserved",
+                "owner_id": request.username,
+            },
+            {
+                "$set": {
+                    "heartbeat": datetime.datetime.utcnow(),
+                }
+            },
+        )
+        return success(dict(updated=results.modified_count > 0))
+
+    def release(self, request: RequestContext):
+        """release a trial"""
+
+        experiment_name = request.data.pop("experiment_name")
+        client = retrieve_experiment_client(request, experiment_name)
+
+        trial_hash = request.data.pop("trial_hash")
+        status = request.data.pop("status")
+        trial = client.get_trial(uid=trial_hash)
+
+        if trial is None:
+            raise ValueError(f"Trial {trial_hash} does not exist in database.")
+
+        client.release(trial, status)
```

Comment:
Sorry, is there a mechanism that handles automatically the error and return them in a json format? I guess I missed it. There are also many potential errors that could occur here: https://github.com/Epistimio/orion/blob/develop/src/orion/client/experiment.py#L492-L499.

Reply:
Yes, it is in the server, all the exceptions are handled. In fact they have to be because the RemoteClient is tested by the regular ExperimentClient unit tests to make sure the coverage is the same and avoid duplicating the tests